Implementing Segmented Video API

christopher-besch opened the pull request:
closes #2096

<!-- Thank you again for contributing! Do not modify the lines below, they are for reviewers. -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

Darylgolden commented:
> 
> 
> Is there a way of stopping those checks from running all the time? This is nowhere near completion.

Probably not easily, but you don't really have to worry about burning GitHub's hardware.

behackl commented:
I've skimmed through your implementation, this looks very good already; great work!

There are two things I've noticed:

- One, and this is in response to your question above, I prefer storing the metadata in a separate json/yaml/whatever file -- but this is mainly because I've had bad experiences with file name parsing in the past. For you, this would break down if users were to put underscores in both the section names and the section type, for example. It also adds a non-obvious limitation to the characters that can be used in these fields as, e.g., Windows has a set of characters that cannot be used in file names.
- And two, the enum for section types is rather particular to presentations and might rather live in the corresponding plugin than the main library, but I also think that this is a minor detail and not as relevant at this point.

Again, well done so far!

behackl commented:
> In addition to that this would be very difficult to implement in a plugin [...]

Sorry, there might be a misunderstanding here (at the very least on my side 😄) -- I think letting Sections have a `type` definitely does make sense, I was not suggestion to remove that part of the implementation. My comment was specifically about the Enum -- I think that the Enum itself could easily live in an external plugin because it belongs there semantically: correct me if I am wrong, but the options currently have no meaning for how the sections are generated; Manim's internal logic does not change depending on the type value (apart from what is written to the metadata).

The one place where you currently use the enum is for setting the type of the first section; we could set the type of it to something generic like `None` and a plugin could special-case this value. (And well, for type hints, the section type could simply be a `str`.)

christopher-besch commented:
> And well, for type hints, the section type could simply be a `str`.

Mh... so far one of my biggest problem with Manim is that type hints don't receive that much love. Maybe we can use some inheritance here, so that the plugin defines a subclass that can be used instead of the default one. The default one would only support `normal`.

I've never worked with the Manim plugin system, is it already mature enough for this?

behackl commented:
As I've mentioned, I don't think it is particularly important that we agree on this; I won't block the PR over it.

We have agreed on "Sections" over "Slides" because the general API should be as flexible and versatile as possible. Including hard-coded, presentation-specific slide types does not fit in this context, in my opinion.

All that I am saying is that in my opinion, specific types should be implemented and live with the corresponding application they are used with. I don't disagree with the proposed types being useful for the presentation application (on the contrary, I think they make a lot of sense), but this PR *is simply not* about the application.

christopher-besch commented:
Okok, you win.
I've implemented an extendible type system based on your idea with `str` and removed the "presentation types". You can take a look at the first comment in this PR, which I've updated.
Is this to your liking?

behackl commented:
Let's merge this to give devs the opportunity to test the changes from this PR on the main branch for a while before the release.

Thank you for your efforts, @christopher-besch!

christopher-besch commented:
And I thank all of you for your continued support. @behackl, @kolibril13, @huguesdevimeux, @Darylgolden and everyone else who kept me motivated!
It wouldn't have been possible without you!
